source/opt/passes.h

```diff
@@ -74,6 +79,138 @@ class EliminateDeadConstantPass : public Pass {
   bool Process(ir::Module*) override;
 };
 
+// The optimization pass for folding SpecConstantOp and SpecConstantComposite
+// instructions whose elements are all known normal constants.
+class FoldSpecConstantOpAndCompositePass : public Pass {
+ public:
+  FoldSpecConstantOpAndCompositePass()
+      : max_id_(0),
+        def_use_mgr_(nullptr),
+        type_mgr_(nullptr),
+        id_to_const_val_() {}
+  const char* name() const override { return "fold-spec-const-op-composite"; }
+  bool Process(ir::Module* module) override {
+    Initialize(module);
+    return ProcessImpl(module);
+  };
+
+ private:
+  // Initializes the type manager, def-use manager and get the maximal id used
+  // in the module.
+  void Initialize(ir::Module* module) {
+    type_mgr_.reset(new analysis::TypeManager(*module));
+    def_use_mgr_.reset(new analysis::DefUseManager(module));
+    for (const auto& id_def : def_use_mgr_->id_to_defs()) {
+      max_id_ = id_def.first > max_id_ ? id_def.first : max_id_;
+    }
+  };
+
+  // The real entry of processing. Iterates through the type, constants, global
+  // variable section of the given module, finds the spec constants defined
+  // with OpSpecConstantOp and OpSpecConstantComposite instructions. If the
+  // result value of those spec constants can be folded, fold them to
+  // corresponding front-end constants.
+  bool ProcessImpl(ir::Module*);
+
+  // Processes the OpSpecConstantOp instruction pointed by the given
+  // instruction iterator, folds it to normal constants if possible. Returns
+  // true if the spec constant is folded to normal constants, otherwise returns
+  // false. The instruction iterator passed in by pointer points to the
+  // OpSpecConstantOp instruction to be processed and also specifies the
+  // instruction insertion position in case new instructions need to be
+  // inserted during the process.
```

Comment:
Since we use pointers to iterators in these methods, it's better to specify what will happen to the iterators after the method call.